src/utils/daos_hdlr.c

```diff
@@ -899,6 +900,41 @@ fs_copy(struct cmd_args_s *ap, struct file_dfs *src_file_dfs, struct file_dfs *d
 		}
 	}
 
+	/* Make sure the source and destination are not the same.
+	 * Make sure the source is not a subset of the destination. */
+	if ((src_file_dfs->type == POSIX && dst_file_dfs->type == POSIX) ||
+	    (src_file_dfs->type == DAOS && dst_file_dfs->type == DAOS &&
+	     strcmp(ca->src_pool, ca->dst_pool) == 0 && strcmp(ca->src_cont, ca->dst_cont) == 0)) {
+		if (strcmp(src_path, dst_path) == 0) {
+			rc = -DER_INVAL;
+			DH_PERROR_DER(ap, rc, "source and destination are the same");
+			D_GOTO(out, rc);
+		}
+		if (S_ISDIR(src_stat.st_mode)) {
+			int  src_path_len;
+			bool src_is_subpath;
+
+			/* Always compare src_path with a trailing slash */
+			src_path_len = strlen(src_path);
+			if (src_path[src_path_len - 1] == '/') {
+				src_is_subpath = (strncmp(src_path, dst_path, src_path_len) == 0);
+			} else {
+				char *tmp_src = NULL;
+
+				D_ASPRINTF(tmp_src, "%s/", src_path);
```

Comment:
should check tmp_src != NULL, but do you really need an extra allocation and a tmp string here?
why not just compare:
(strncmp(src_path, dst_path, src_path_len) == 0 && dst_path[src_path_len] == "/") ?


im not sure though if this is correct though, because consider
what if src == dst == /d1/d2
I see that the same as
src == /d1/d2 and dst == /d1/d2/
but the result is different in this case, no?

Reply:
Alternatively you could stat the src and dest and if the `st_dev` and `st_ino` fields are the same then they two are the same.  This has the advantage that it'll work if you access the same tree via different paths, full vs relative or in the presence of symlinks etc.

Reply:
> why not just compare:
(strncmp(src_path, dst_path, src_path_len) == 0 && dst_path[src_path_len] == "/") ?

You're right. I tested locally and that works the same (with minor tweaks).

Per my other comment to @wiliamhuang: https://github.com/daos-stack/daos/pull/13962#discussion_r1567945589

At this point, `dst_path` is either `/`, has a `/` at the same location as `src_path`, or `src_path` is not a subpath of `dst_path`.


For my previous implementation and your suggestion, all four of these error as expected:
```
daos fs copy --src src  --dst src
daos fs copy --src src  --dst src/
daos fs copy --src src/ --dst src/
daos fs copy --src src/ --dst src
```

I pushed your suggestion and added more comments here

Reply:
> Alternatively you could stat the src and dest and if the st_dev and st_ino fields are the same

That would be more complete, but I think we'd have to do some recursive parse + stat to detect subpaths. E.g.
```
src_path = /some/deep/dir
dst_path = /some/deep/dir/that/keeps/going
```
We'd have to stat like this to figure it out (unless I'm missing something)
```
src - /some/deep/dir
dst - /some/deep/dir/that/keeps/going
dst - /some/deep/dir/that/keeps
dst - /some/deep/dir/that
dst - /some/deep/dir
```